File: 1297-maximum-number-of-balloons/1297-maximum-number-of-balloons.java
class Solution {
    public int maxNumberOfBalloons(String text) {
        int b = 0, a = 0, l = 0, o = 0, n = 0;
        for (char aChar : text.toCharArray())
            switch (aChar) {
                case 'b' : b++;
                break;
                case 'a' : a++;
                break;
                case 'l' : l++;
                break;
                case 'o' : o++;
                break;
                case 'n' : n++;
                break;
            }
        return Math.min(Math.min(Math.min(b,a),n), Math.min(l/2,o/2));
        
    }
}

/*
class Solution {
    public int maxNumberOfBalloons(String text) {
        int n = text.length();
        
        Map<Character, Integer> mp = new HashMap<>();
        StringBuilder temp = new StringBuilder("balloon");

        for(int i = 0; i < temp.length(); i++){
            mp.put(temp.charAt(i),0);
        }

        for(char c : text.toCharArray()){
            mp.put(c, mp.getOrDefault(c,0)+1);
        }

        mp.put('l',mp.getOrDefault('l',0)/2);
        mp.put('o',mp.getOrDefault('o',0)/2);

        int ans = Integer.MAX_VALUE;

        ans = Math.min(ans, mp.get('b'));
        ans = Math.min(ans, mp.get('a'));
        ans = Math.min(ans, mp.get('l'));
        ans = Math.min(ans, mp.get('o'));
        ans = Math.min(ans, mp.get('n'));

        return ans;
    }
}
*/

File: 1297-maximum-number-of-balloons/1297-maximum-number-of-balloons.cpp
class Solution {
public:
    int maxNumberOfBalloons(string text) {
        int n = text.size();
        string s = "ballon";

        unordered_map<char, int> mp;
        for(auto x : s){
            mp[x] = 0;
        }

        for(auto x : text){
            if(mp.count(x)){
                mp[x]++;
                //ans = min(ans, mp[x]);
            }
        }

        // Since 'l' and 'o' appear twice in "balloon", we divide their counts by 2
        mp['l'] /= 2;
        mp['o'] /= 2;

        // The number of "balloon"s we can form is determined by the minimum count among the characters
        int ans = INT_MAX;
        ans = min(ans, mp['b']);
        ans = min(ans, mp['a']);
        ans = min(ans, mp['l']);
        ans = min(ans, mp['o']);
        ans = min(ans, mp['n']);

        return ans;
    }
};



/*
class Solution {
public:
    int maxNumberOfBalloons(string text) {
        int n = text.size();
        string s = "ballon";

        unordered_map<char, int> mp;
        for(auto x : s){
            mp[x] = 0;
        }

        int ans = INT_MAX;

        for(auto x : text){
            if(mp.count(x)){
                mp[x]++;
                //ans = min(ans, mp[x]);
            }
        }

        for(auto x : mp){
            ans = min(x.second, ans);
            if(x.first == 'l' && x.second == 1){
                ans = 0;
            }
            if(x.first == 'o' && x.second == 1){
                ans = 0;
            }
        }

        return ans;
    }
};

*/